File: risk-lib/src/test/java/net/mjduffin/risk/lib/usecase/DraftTests.java
package net.mjduffin.risk.lib.usecase;

import net.mjduffin.risk.lib.entities.*;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.*;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

class DraftTests {
    private final String BOB = "Bob";
    private final String ALICE = "Alice";

    PlayerInput getPlayerInputFromGame(Game game) {
        DieThrow dieThrow = Mockito.mock(DieThrow.class);
        when(dieThrow.getDieValue()).thenReturn(1);
        DiceManager diceManager = new DiceManager(dieThrow);
        return new GameManager(game, diceManager);
    }

    @Test
    void draft() throws GameplayException {
        // given
        Game.Builder gameBuilder = new Game.Builder();

        gameBuilder.addPlayerWithTerritories(BOB, Arrays.asList("England", "Wales"));
        Game game = gameBuilder.build();

        PlayerInput playerInput = getPlayerInputFromGame(game);

        // when
        Map<String, Integer> draft = new HashMap<>();
        draft.put("England", 1);
        draft.put("Wales", 3);
        playerInput.draft(BOB, draft);

        assertEquals(2, game.getBoard().getTerritory("England").getUnits());
        assertEquals(4, game.getBoard().getTerritory("Wales").getUnits());
    }

    @Test
    void draftTwiceInRow() throws GameplayException {
        // given
        Game.Builder gameBuilder = new Game.Builder();

        // TODO: This is currently dependent on the order
        gameBuilder.addPlayerWithTerritories(BOB, Collections.singletonList("Wales"));
        gameBuilder.addPlayerWithTerritories(ALICE, Collections.singletonList("England"));
        Game game = gameBuilder.build();

        PlayerInput playerInput = getPlayerInputFromGame(game);

        Map<String, Integer> draft = new HashMap<>();
        draft.put("Wales", 4);
        playerInput.draft(BOB, draft);
        draft.clear();
        draft.put("England", 1);
        playerInput.draft(ALICE, draft);
        assertThrows(GameplayException.class, () -> playerInput.draft(BOB, draft));

    }

}

File: risk-lib/src/test/kotlin/net/mjduffin/risk/lib/usecase/DraftTests.kt
package net.mjduffin.risk.lib.usecase

import net.mjduffin.risk.lib.entities.Board
import net.mjduffin.risk.lib.entities.DiceManager
import net.mjduffin.risk.lib.entities.DieThrow
import net.mjduffin.risk.lib.entities.Game
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

internal class DraftTests {
    private val BOB = "Bob"
    private val ALICE = "Alice"

    private fun getGameManagerFromGame(game: Game): GameManager {
        val dieThrow = Mockito.mock(DieThrow::class.java)
        Mockito.`when`(dieThrow.dieValue).thenReturn(1)
        val diceManager = DiceManager(dieThrow)
        val board = Board.Builder().addEdge("England", "Wales").build()
        return GameManager(board, game, diceManager)
    }

    @Test
    fun draft() {
        // given
        val gameBuilder = Game.Builder()
        gameBuilder.addPlayerWithTerritories(BOB, listOf("England", "Wales"))
//        gameBuilder.addEdge("England", "Wales")
        val game = gameBuilder.build()
        val gameManager = getGameManagerFromGame(game)

        // when
        val draft: MutableMap<String, Int> = HashMap()
        draft["England"] = 1
        draft["Wales"] = 3
        gameManager.draft(BOB, draft)
        val gameState = gameManager.getGameState()

        // then
        val expected = GameState(BOB, "END", 6, listOf("England", "Wales"), listOf(BOB, BOB), listOf(2, 4))
        assertEquals(expected, gameState)
    }

    @Test
    fun draftTwiceInRow() {
        // given
        val gameBuilder = Game.Builder()

        // TODO: This is currently dependent on the order
        // when
        gameBuilder.addPlayerWithTerritories(BOB, listOf("Wales"))
        gameBuilder.addPlayerWithTerritories(ALICE, listOf("England"))
        val game = gameBuilder.build()
        val playerInput = getGameManagerFromGame(game)
        playerInput.draft(BOB, mapOf(Pair("Wales", 4)))
        playerInput.draft(ALICE, mapOf(Pair("England", 1)))
        // then
        assertFailsWith<GameplayException> { playerInput.draft(BOB, mapOf(Pair("England", 1))) }
    }

    @Test
    fun `can draft in multiple stages`() {
        // given
        val gameBuilder = Game.Builder()

        // TODO: This is currently dependent on the order
        // when
        gameBuilder.addPlayerWithTerritories(BOB, listOf("England"))
        gameBuilder.addPlayerWithTerritories(ALICE, listOf("Wales"))
        val game = gameBuilder.build()
        val playerInput = getGameManagerFromGame(game)

        // then
        playerInput.draftSingle(BOB, "England", 3)
        playerInput.draftSingle(BOB, "England", 3)
        val gameState = playerInput.getGameState()
        assertEquals("ALLDRAFT", gameState.phase)
        assertEquals(listOf("England", "Wales"), gameState.territories)
        assertEquals(listOf(BOB, ALICE), gameState.occupyingPlayers)
        assertEquals(listOf(7, 1), gameState.units)

        // should not be able to end turn until finished drafting
        assertFailsWith<GameplayException> { playerInput.endTurn(BOB) }

        // finish drafting
        playerInput.draftSingle(BOB, "England", 4)
        assertEquals(ALICE, playerInput.getGameState().currentPlayer)
    }
}